pruner/pruner.go

```diff
@@ -0,0 +1,339 @@
+	}
+	if m.rebuildTrigger.Rebuild(added, &m.stats) {
+		return m.rebuild(added)
+	}
+
+	return nil
+}
+
+// AddPattern calls the underlying AddPattern method and then maybe
+// rebuilds the index (if the AddPattern succeeded).
+func (m *Matcher) AddPattern(x quamina.X, pat string) error {
+	var err error
+
+	// Do we m.live.Add first, do we m.Matcher.AddPattern first?
+	if err = m.Matcher.AddPattern(x, pat); err == nil {
+		m.lock.Lock()
+		m.stats.Added++
+		m.stats.Live++
+		m.maybeRebuild(true)
+		m.lock.Unlock()
+		err = m.live.Add(x, pat)
+		// ToDo: Contemplate what do to about an error here
+		// (or if we got an error from AddPattern after we did
+		// live.Add.
+	}
+
+	return err
+}
+
+// NewFJ just calls quamina.FJ.
+func NewFJ(m *Matcher) quamina.Flattener {
+	return quamina.NewFJ(m.Matcher)
+}
+
+// NewFJ calls quamina.NewFJ with this Matcher's core quamina.Matcher
+func (m *Matcher) NewFJ() quamina.Flattener {
+	return quamina.NewFJ(m.Matcher)
+}
+
+// MatchesForJSONEvent calls the underlying MatchesForJSONEvent method
+// and then maybe rebuilds the index.
+func (m *Matcher) MatchesForJSONEvent(event []byte) ([]quamina.X, error) {
+	fs, err := m.NewFJ().Flatten(event)
+	if err != nil {
+		return nil, err
+	}
+	return m.MatchesForFields(fs)
+}
+
+func (m *Matcher) MatchesForFields(fields []quamina.Field) ([]quamina.X, error) {
+
+	xs := m.Matcher.MatchesForFields(fields)
+
+	// Remove any X that isn't in the live set.
+
+	acc := make([]quamina.X, 0, len(xs))
+
+	// We're updating stats.Filtered, so we need a write lock.  If
+	// we forget about stats.Filtered, we can live with only a
+	// read lock here.
+	var emitted, filtered int64
+	for _, x := range xs {
+		have, err := m.live.Get(x)
+		if err != nil {
+			return nil, err
+		}
+		if have == "" {
+			filtered++
+			continue
+		}
+		acc = append(acc, x)
+		emitted++
+	}
+
+	m.lock.Lock()
+	m.stats.Filtered += filtered
+	m.stats.Emitted += emitted
+	m.maybeRebuild(false)
+	m.lock.Unlock()
+
+	return acc, nil
+}
+
+// DelPattern "removes" the pattern from the index and maybe rebuilds
+// the index.
+func (m *Matcher) DelPattern(x quamina.X) (bool, error) {
+	had, err := m.live.Del(x)
+	if err == nil {
+		if had {
+			m.lock.Lock()
+			m.stats.Deleted++
+			m.stats.Live--
+			m.maybeRebuild(false)
+			m.lock.Unlock()
+		}
+	}
+
+	return had, err
+}
+
+// Rebuild rebuilds the matcher state based on only live patterns.
+//
+// If calling fearlessly, then the old matcher is released before
+// building the new one.
+//
+// This method resets the Stats.
+func (m *Matcher) Rebuild(fearlessly bool) error {
+	m.lock.Lock()
+	err := m.rebuild(fearlessly)
+	m.lock.Unlock()
+	return err
+}
+
+// rebuild is Rebuild but assumes having the lock.
+func (m *Matcher) rebuild(fearlessly bool) error {
+	// We assume we have the lock.
+
+	// Nothing fancy here now.
+
+	var (
+		then = time.Now()
+		m1   = quamina.NewMatcher()
+	)
+
+	if fearlessly {
+		// Let the GC reduce heap requirements?
```

Comment:
I see, the idea is not to have two Matchers in memory. On the other hand, we worry about a failed rebuild() destroying a working Matcher.   We'll know which is the right thing to worry about after some more experience...